### ISSUE TITLE
Once the invitation link is clicked, even if the form is not submitted, a new click leads to a 404

### ISSUE THREAD

**Issue by tanzaho** [bee-keeper/django-invitations#25]
Hi,
I am using allauth. The integration with django-invitations works perfectly. 

However, I encounter a behaviour that seems strange to me. Once an invitation link is clicked, the invitation is 'accepted', even if the user doesn't submit a sign up form. So, if a user inadvertently closes the page and wants to reload it, it will find a 404 error.

Another related behaviour that I find strange, is the fact that a 404 is raised for previously "accepted" invites. I would like instead to display an error message like "Sorry this invitation has expired" or "Sorry this invitation has already been used"..

**Comment by bee-keeper** [comment]
Hi there,

Thanks for raising these points.

I suppose in the first case, it would make more sense to keep the invite as unaccepted until either the user signs up or the invite expires.  This would have to be specific to the allauth backend though.

Re your second point you're right, a straight 404 is not the way to go.  One idea I came up with was to redirect to a configured url and use the messages framework to explain the invitation has expired/been  accepted.  The default behaviour could be simply 410 for link gone.  I think this might need some more thinking through though.

In either case I'd accept a PR to fix this behaviour.

**Comment by tanzaho** [comment]
I'll be working on it. I'll let you know.

**Comment by tanzaho** [comment]
About the first point, why is it only related to allauth ? I haven't tried to use invitations without it, but I understand from the code that the behaviour would be similar : one click on the link leads to a given url -> invitations is accepted -> further clicks would lead to a 404 or any other error message).

However, by thinking like this, it means that one would need a signal coming from the target URL, which of course depends on any arbitrary module.

**Comment by bee-keeper** [comment]
So firstly, I've just cut and pushed a new devel branch for these changes.

It's only related to allauth in that the default without using allauth is simply the user clicks a link and is redirected to a preconfigured url which might not necessarily be a signup form.  I think your use case is common enough to be handled but I think it's important to remember redirecting to a signup form is only 1 potential use case, even if using the allauth backend.

Given these complications I think the expired invitation message is probably the easier problem to tackle here.

**Comment by tanzaho** [comment]
I completely understand. I'll probably have something by tomorrow for that.

**Comment by bee-keeper** [comment]
I'm leaving this issue open as your primary concern (below) has still not been addressed.  I suspect we need a boolean flag on each invitation and an app setting allowing users to decide how they want to define this behaviour.  I'm going to get another release out and I'll come back to this.

> However, I encounter a behaviour that seems strange to me. Once an invitation link is clicked, the invitation is 'accepted', even if the user doesn't submit a sign up form. So, if a user inadvertently closes the page and wants to reload it, it will find a 404 error.

**Comment by clokep** [comment]
@bee-keeper I'm wonder if something like this would work:

1. The view calls a method on the adapter to handle the invitation being 'accepted'.
2. The allauth adapter stashes this information in the session.
3. The allauth adapter hooks into the signal that a user has registered. It then checks the session to see if that user had previously accepted an invite.

I don't know what to do for the non-allauth adapter. :) Maybe use the login signal instead of the register signal? But that has the obvious issue of if someone logs in with a different account.

**Comment by nsomaru** [comment]
Hello,

@clokep have you made any progress on this?

I would like to help to resolve this issue if possible. @bee-keeper -- could you recommend a strategy for resolving and mentor me through a patch?

**Comment by clokep** [comment]
@nsomaru I hacked around this locally, I subclassed the [`AcceptInvite` view](https://github.com/bee-keeper/django-invitations/blob/3ed7baa02fc142864670ef497fef4872d8e89240/invitations/views.py#L90-L168), overrode the `post` method to remove [two](https://github.com/bee-keeper/django-invitations/blob/3ed7baa02fc142864670ef497fef4872d8e89240/invitations/views.py#L141-L143) sets [of lines](https://github.com/bee-keeper/django-invitations/blob/3ed7baa02fc142864670ef497fef4872d8e89240/invitations/views.py#L147-L149). I then hooked into the `user_signed_up` from django-allauth with something like:

```python
@receiver(user_signed_up)
def accept_invite(sender, request, user, **kwargs):
    # Traverse from the user to verified email addresses. It is possible for a
    # user to already have multiple email addresses if they typed in a different
    # email after accepted the invite. In this case, the user will have two
    # emails, but only one will be verified.
    addresses = EmailAddress.objects.filter(user=user, verified=True) \
                                    .values_list('email', flat=True)
    # Check if any invites exist for this address and were accepted.
    invites = Invitation.objects.filter(email__in=addresses)
    if invites:
        # Mark all these invites as accepted.
        invites.update(accepted=True)
        for invite in invites:
            # Note that this doesn't send it with a request.
            invite_accepted.send(sender=Invitation, email=invite.email)

        # Figure out if you care if there are multiple invites here.
```

This is essentially what I proposed [above](https://github.com/bee-keeper/django-invitations/issues/25#issuecomment-182979386). I'm still willing to take a look at this (or help review some code from @nsomaru), but I'm unsure how to handle the non-allauth adapter.

Note that I've been running with this in production for a handful of months now.

**Comment by bee-keeper** [comment]
@clokep thanks for your suggestion, and sorry for taking so long to respond.  Please feel free to put together a PR and we can take it from there.
 
@nsomaru feel free to get involved in the PR

The basic adapter is more troublesome but at least we can get this fixed to the allauth adapter.

**Comment by erm** [comment]
Just wanted to mention that this was an issue that forced me to abandon django-invitations for my current project. I've forked the repo and will attempt to revisit it later and put in a PR myself if someone doesn't beat me to it - however I don't have enough time to do so now and have a deadline to meet.

**Comment by nsomaru** [comment]
@erm apologies, this should have been posted earlier. [Here's a workaround](https://gist.github.com/nsomaru/53d0caf0981b56d57d77649b29c8cd1d) from @clokep's solution above.

**Comment by williamwang61** [comment]
@nsomaru Hi, I saw your workaround of the issue that invitation is only accepted after the user finished signup, which is exactly how I did on my local. Since this is a common problem, do you plan to make a PR for it? Otherwise I can make it. Thanks!

**Comment by nsomaru** [comment]
@williamwang61 as per this thread we would need to handle the non allauth case for the PR to be accepted. I suppose adding a setting (like `INVITATIONS_COMPLETED_FIELD`) to point to the `(EmailAddressKlass, field_name)` you want to check? @bee-keeper care to weigh in?

**Comment by bee-keeper** [comment]
Hi, sorry I haven't had much time to work on invitations recently.  

Please do submit a PR though for the allauth adapter, we can work out the details for the non auth adapter at a later stage.  It seems like this is a popular feature so would be good to get it in a release asap.

**Comment by williamwang61** [comment]
@bee-keeper Hey, I just created a PR for that. I'd appreciate it if you could have a look. #https://github.com/bee-keeper/django-invitations/pull/49

**Comment by bee-keeper** [comment]
Thanks @williamwang61 and @clokep - have just merged and released 1.8.

**Comment by BasemSrour** [comment]
Hi can you help me with this error, I use all auth with django-invitations too but i can't solve this error.
NotImplementedError: Unable to find a unique username.
